src/javaServerStarter.ts

```diff
@@ -1,24 +1,102 @@
   options.env = process.env;
   options.stdio = 'pipe';
   executable.options = options;
-  executable.command = path.resolve(requirements.java_home + '/bin/java');
-  executable.args = prepareParams(requirements, xmlJavaExtensions, context);
-  return executable;
+  const useBinary = getXMLConfiguration().get("server.binary.enabled", false);
+  const binaryOptions: string = getXMLConfiguration().get("server.binary.args");
+
+  /**
+   * Finish preparing the arguments for the java xml language server and return it
+   *
+   * @returns The configured java xml language server
+   */
+  const javaExecutable: () => Executable = (): Executable => {
+    executable.command = path.resolve(requirements.java_home + '/bin/java');
+    executable.args = prepareParams(requirements, xmlJavaExtensions, context);
+    return executable;
+  }
+
+  /**
+   * Attempt to use the binary executable, but fall back to java if something goes wrong
+   */
+  const binaryExecutable: () => Thenable<Executable> = (): Thenable<Executable> => {
+    return getServerBinary().then((serverBinary: string): Thenable<Executable> => {
+
+      /**
+       * Finish preparing the arguments for the binary xml language server and return it
+       *
+       * @returns The configured binary xml language server
+       */
+      const finalizeBinary: () => Executable = (): Executable => {
+        executable.args = [binaryOptions];
+        executable.command = serverBinary;
+        return executable;
+      };
+
+      const hash: Hash = createHash('sha256');
+      hash.update(fs.readFileSync(path.resolve(serverBinary)));
+      const hashDigest: string = hash.digest('hex').toLowerCase();
+      const expectedHashPath: string = path.resolve(__dirname, '../server', `lemminx-${os.platform}-hash`);
+      try {
+        const expectedHash = fs.readFileSync(expectedHashPath).toString('utf-8').toLowerCase().split(' ')[0];
+        if (hashDigest === expectedHash) {
+          return Promise.resolve(finalizeBinary());
+        }
+        return window.showErrorMessage(`The server binary ${serverBinary} has been modified! `
+        + 'Running the file poses a threat to your system\'s security. '
+        + 'Run anyways?', 'Yes', 'No').then((val: string): Executable => {
+          if (val === 'Yes') {
+            return finalizeBinary();
+          }
+          return javaExecutable();
+        });
+      } catch (err) {
+        return window.showErrorMessage('The integrity of the server binary cannot be verified, since the hash is missing. '
+          + 'Running the file poses a threat to your system\'s security. '
+          + 'Run anyways?', 'Yes', 'No').then((response: string): Executable => {
+          if (response === 'Yes') {
+            return finalizeBinary();
+          }
+          return javaExecutable();
+        });
+      }
+    }, (err): Promise<Executable> => {
+      window.showErrorMessage('Failed to download the binary server. Falling back to the Java server...');
```

Comment:
Failed to download the binary XML server. Falling back to the Java-based XML server...